Build with Visual Studio 2015 on Windows

gqmelo opened the pull request:
While trying to build Mesa's llvmpipe on Windows, I've got lots of link errors like ```LLVMCodeGen.lib(LowerEmuTLS.obj) : error LNK2038: mismatch detected for '_MSC_VER': value '1800' doesn't match value '1900'``` and then realized that even though ```llvmdev``` declares ```vc14``` it is actually compiled with VS2013.

I believe this was done by mistake, but in case a VS2013 build is still really necessary, we should not generate a package with ```vc14``` and probably depend on ```vs2013_runtime```

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

gqmelo commented:
We should probably rebuild https://github.com/conda-forge/clangdev-feedstock too

SylvainCorlay commented:
(FYI @gqmelo this now conflicts with master.)

I would be very much in favor of switching to Visual Studio 2015. Indeed this will require to rebuild `clangdev`.

@inducer was there a particular reason to use an older version of msvc?

gqmelo commented:
Thanks, rebased it.

SylvainCorlay commented:
I think that after #10 , the fix is simply going to remove the first lines of the bld.bat which call msvcvarsall.bat with msvc2013.

Allso, we should probably add the msvc14 C runtime as a runtime dependency on windows.

gqmelo commented:
> Allso, we should probably add the msvc14 C runtime as a runtime dependency on windows.
Thanks, I've added it before but lost it during the first rebase.

> I think that after #10 , the fix is simply going to remove the first lines of the bld.bat which call msvcvarsall.bat with msvc2013.

I rebased it again. But I've also reverted to not use nmake. nmake is not a good option as it does not support parallel compilation.

SylvainCorlay commented:
> I rebased it again. But I've also reverted to not use nmake. nmake is not a good option as it does not support parallel compilation.

#11 also switches to msvc15. I prefer nmake over vcproj files since it makes the  build script much more similar to that of unix systems.

It is my understanding that nmake can use multiple cores for building if one sets `set CL=/MP`. Although I have not tried.

SylvainCorlay commented:
I don't know why the library was originally built with an older version of msvc.

gqmelo commented:
> It is my understanding that nmake can use multiple cores for building if one sets set CL=/MP. Although I have not tried.

The problem is that you normally need to modify the generated build scripts:
http://stackoverflow.com/a/601984

Another option would be to use ninja (available on conda-forge), which will work like ```make```. Don't you want to try on https://github.com/conda-forge/llvmdev-feedstock/pull/11 ? It is just a matter of adding ninja as build dependency, pass ```-GNinja``` to cmake and call ```ninja``` and ```ninja install```
I think it is nice to have a parallel build by default, specially when trying to build locally to debug something.

gqmelo commented:
FYI, compare the duration of https://ci.appveyor.com/project/conda-forge/llvmdev-feedstock/build/1.0.48 and https://ci.appveyor.com/project/conda-forge/llvmdev-feedstock/build/1.0.45

nmake duration is more 2x

SylvainCorlay commented:
I have no experience with ninja.

The advantage of nmake over vcproj is that things are set up through the environment while vcproj are specific to a version of msvc.

If there was an equivalent ninja build method, which was also faster, I would be very much in favor of it.

SylvainCorlay commented:
Now that the week-end is over, my main worry with the update to msvc 2015 is that we need to build llvm first and clang after.

With the size of the os-x backlogs during the week, this will create a gap of multiple days during which the llvm version is not compatible with clang.

For this reason, unless the core conda-forge folks have a means to prevent this gap, I would like to wait for the low week-end backlog to merge the switch to msvc 2015.

gqmelo commented:
> The advantage of nmake over vcproj is that things are set up through the environment while vcproj are specific to a version of msvc.

I'm not sure what you mean with "set up through the environment", but for ninja, cmake generates scripts to be used with ninja, just like ```make```, ```nmake```, etc. It's pretty straightforward.

SylvainCorlay commented:
> I'm not sure what you mean with "set up through the environment".

through environment variables set by `vcvarsall.bat`.

> cmake generates scripts to be used with ninja, just like make, nmake, etc. It's pretty straightforward.

sounds good.

SylvainCorlay commented:
> Now that the week-end is over, my main worry with the update to msvc 2015 is that we need to build llvm first and clang after.
>
> With the size of the os-x backlogs during the week, this will create a gap of multiple days during which the llvm version is not compatible with clang.
>
> For this reason, unless @conda-forge/core folks have a means to prevent this gap, I would like to wait for the low week-end backlog to merge the switch to msvc 2015.

Or we can just go ahead and merge #11. Then rebuild https://github.com/conda-forge/clangdev-feedstock/pull/5 and merge it.

gqmelo commented:
> through environment variables set by vcvarsall.bat

Ok, then yes, ninja will just get the compiler set by ```vcvarsall.bat```

gqmelo commented:
> For this reason, unless the core conda-forge folks have a means to prevent this gap, I would like to wait for the low week-end backlog to merge the switch to msvc 2015.

Unfortunately I think there is no mechanism to avoid that. I'm ok if you want to wait until the weekend. I can wait for https://github.com/conda-forge/staged-recipes/pull/2330 until there

SylvainCorlay commented:
> > For this reason, unless the core conda-forge folks have a means to prevent this gap, I would like to wait for the low week-end backlog to merge the switch to msvc 2015.
>
> Unfortunately I think there is no mechanism to avoid that. I'm ok if you want to wait until the weekend. I can wait for conda-forge/staged-recipes#2330 until there

I also need an msvc 2015 build for building cling. I would be in favor of going ahead with the merge and merge the clang msvc2015 build as soon as possible.

gqmelo commented:
Ok then, your call. But if you know someone who is using ```clangdev``` for sure it would be nice to leave a message telling to use llvmdev build number 2 for now.

gqmelo commented:
I'm closing this PR in favour of https://github.com/conda-forge/llvmdev-feedstock/pull/11
